### ISSUE TITLE
provide scan ignore specification

### ISSUE THREAD

**Issue by timowolf** [fossology/fossology#703]
We upload complete repository for scanning. The repos often contain test code, documentation, or demo code. This code is not included in the system at runtime and not shipped in the application. From time to time, this code includes other licenses or copyrights. At the moment that scanning result includes this licenses and copyrights.

From a developer view, I would like to specify folders and files like in a .gitignore file, which are not relevant for the runtime or shipping of my application or component. This would reduce the manual rework of the scan results.

This could become a generic spec, that is used to define which source files of a repo are not part of the runtime, shipped component or application.

What do you think about a .scanignore, working the same way as a .gitignore?

**Comment by rlintu** [comment]
You may already know that you can ignore scanning results by selecting 'Make decisions irrelevant'.
I'm using that all the time. Of course if you are scanning the same build tree over and over again, then
there is some duplicate work.

**Comment by timowolf** [comment]
Thanks for the notice. This is new to me, as I do not use fossology. Our organization has a team doing it. This is also the reason for this request. When I, as a developer, know upfront what files are no relevant for fossology, it would be great to mark them within an ignore file. I believe this will reduce lots of manual post scanning work and communication across teams.

**Comment by rlintu** [comment]
Ok I got your point. Because the analysis is always done for the snapshot I would suggest as an interim solution that you build for your team a clean package for analysis using 
````tar --exclude-vcs -X file ...````
which only has the relevant part of your code and which is then uploaded to FOSSology. Actually I'm using that approach and I have scripts which I can reuse for building packages for analysis.

Anyway your suggestion is something to consider in the future.

**Comment by timowolf** [comment]
Thanks for the hint and for the short-term solution. This will help for the current work. But I envision here general long-term solution, which could be independent of fossology.

We are mainly using git as an scm system. When providing code for clearing I specify the tag in git and download the archive of the repository. Than, I get a great bundle with version id and hash. This is uploaded to fossology and we have full traceability. I also do the same when clearing open source projects from for example github. Changing the archive, I downloaded from github (applied for any other repository too), seems to me faulty.

Should be part of the OSS project to define these .scanignores and not my job. The archive to scan should exactly be a tagged version in the scm system. And when two different scans from different organizations scan the same OSS component, the result should be the same.

Therefore, I do not like a manual change step of code before sending to fossology.

**Comment by mcjaeger** [comment]
.scanignore sounds like a good plan. Point is that for some compilers or MDA generation infrastructure (which do not go into distribution) they do actually have consequences in licensing. The person who does that scan ignore should do a super careful job then. But if the scan ignore is actually a good idea when it is part of the OSS project itself (and thus reviewed).

**Comment by bufferoverflow** [comment]
`.scanignore` is probably to generic, what about a `.licenseignore` ?
Whis would be useful when we start to scan fossology with fossology to exclude folders such as `src/nomos/agent_tests/testdata`.

**Comment by timowolf** [comment]
At the end it is just a word. But .licenseignore implies that I want to ignore a license. And this is definitely not intended. What I actually want to have, is a spec of files that are definitely not in the runtime. How about .notinruntime ?

This file should definitely be in the SCM system of the oss project. The goal is to have identical scan results even when two different organizations do the scanning and with no manual manipulation of the archived repository file. 

There could also be a fossology flag to scan all files and ignoring this spec file.

**Comment by bufferoverflow** [comment]
@timowolf Yes, I'm on the same page.. ignore licenses is not what we need. Scratch that proposal!

I just had a look at the package.json spec and there is a `.npmignore` or the files definition within the `package.json`. An option would be that fossology detects the package type, e.g. node.js and scans accordingly. I'm pretty sure other package definitions provide similar options.

**Comment by NicolasToussaint** [comment]
A .scanignore feature would definitely be useful.

Also, specifically concerning SCM data (.git and so on), we could add an option to "ignore all SCM related files" when uploading a new project to be scanned.
Or maybe even always ignore those files without any option ?

**Comment by atgreen** [comment]
This is an important and useful feature.  To be useful, the scanner needs to distinguish between runtime code and build/test-time code.  My project, libffi, has a very liberal runtime license, but is bundled with GPL build and test scripts.  When fossology users reach out to me for clarification, my go-to response is to tell them to stop using fossology because it is not helpful.  I don't mind annotating my project to distinguish between build- and run-time source (in a .scanignore or similar) if it will help.

**Comment by NicolasToussaint** [comment]
Has anyone worked on this ?

**Comment by NicolasToussaint** [comment]
Hi, what would you think about accepting a new `.fossology-do-not-use` file, containing a list of files and folders, that would automatically and recursively be marked as "Do Not Use" (see #1560)

We could start with a very simple file, accepting one absolute path per line, no regex (nothing fancy like `.git-ignore` files)

Of course, as @mcjaeger mentioned, this file has a important impact on the final clearing decisions, so extra care should be brought when redacting the file. Maybe it should be possible to [de]activate the feature for new uploads.

Technically, I'm not sure ; maybe the simplest would be to have the decider agent apply the decision ?

**Comment by timowolf** [comment]
I like it. I would propose to also be able to address complete folders like a docs folder and I would propose another name. I think .fossology-do-not-use is very long and specific.

How about: .oss-scan-ignore

**Comment by NicolasToussaint** [comment]
@timowolf yes, I also had in mind to allow both files and folder paths.

Concerning the file name, I guess it all depends if we intend to work on seeing this feature implemented in other scanners or not - I personally do not have a strong opinion on which filename is best ; let's see what others in the community think ?

**Comment by NicolasToussaint** [comment]
I would be very interested to know if that feature still sounds interesting to Fossology users here !

If yes, looks like the only point left is the name of the file...  
For this, we could 
- make this configurable
- accept multiple filenames
WDYT ?

**Comment by timowolf** [comment]
@NicolasToussaint how about .runtimeignore

**Comment by shaheemazmalmmd** [comment]
Related to https://github.com/fossology/fossology/issues/2848 . we don't need 2 issues so closing this one.
